---
title: "R Reviw Project"
output: html_notebook
---

***Downloading TCGA-SKCM Data***
-> code from R data loading
```{r}
knitr::opts_knit$set(root.dir = normalizePath("/Users/mi/Desktop/QBIO/qbio_490_jeanneR/analysis_data"))

if (!require("BiocManager", quietly = TRUE)) install.packages("BiocManager")
BiocManager::install(version = "3.17")
if (!require("TCGAbiolinks", quietly = TRUE)) BiocManager::install("TCGAbiolinks")
if (!require("maftools", quietly = TRUE)) BiocManager::install("maftools")
library(BiocManager) 
library(TCGAbiolinks) 
library(maftools)

clin_query <- GDCquery(project = "TCGA-SKCM", data.category = "Clinical",
data.type = "Clinical Supplement", data.format = 'BCR Biotab')
GDCdownload(clin_query)
clinical.BCRtab.all <- GDCprepare(clin_query)
clinic <- clinical.BCRtab.all$clinical_patient_skcm[-c(1,2),]

colnames(clinic)[ colnames(clinic) == "bcr_patient_barcode" ] <- "Tumor_Sample_Barcode"

write.csv(clinic, "/Users/mi/Desktop/QBIO/qbio_490_jeanneR/analysis_data/skcm_clinical_data.csv", row.names=FALSE)

clinic <- read.csv("/Users/mi/Desktop/QBIO/qbio_490_jeanneR/analysis_data/skcm_clinical_data.csv")

rad <- clinical.BCRtab.all$clinical_radiation_skcm[-c(1,2),]
drug <- clinical.BCRtab.all$clinical_drug_skcm[-c(1,2),]

```


```{r}
maf_query <- GDCquery(project = "TCGA-SKCM",
                      data.category = "Simple Nucleotide Variation",
                      access = "open",
                      data.type = "Masked Somatic Mutation",
                      workflow.type = "Aliquot Ensemble Somatic Variant                         Merging and Masking")
GDCdownload(maf_query)
maf <- GDCprepare(maf_query)

rna_clinical$barcode <- rna_clinical$patient
names(rna_clinical)[names(rna_clinical) == 'barcode'] <- 'Tumor_Sample_Barcode'

maf_object <- read.maf(maf = maf, clinicalData = rna_clinical, isTCGA = TRUE)
```

```{r}
rna_query <- GDCquery(project ="TCGA-SKCM",
                      data.category = "Transcriptome Profiling",
                      data.type = "Gene Expression Quantification",
                      workflow.type = "STAR - Counts") 
GDCdownload(rna_query)
rna_se <- GDCprepare(rna_query)

rna_clinical <- rna_se@colData
rna_clinical <- as.data.frame(rna_clinical)

rna_genes <- rna_se@rowRanges@elementMetadata
rna_genes <- as.data.frame(rna_genes)

rna_counts <- rna_se@assays@data$unstranded
rna_counts <- as.data.frame(rna_counts)


```

***Making a KM Plot***
-> code from Intro to Clinical
```{r}
if (!require(survival)) {
  install.packages("survival")
}

if (!require(survminer)) {
  install.packages("survminer")
}

if (!require(ggplot2)) {
  install.packages("ggplot2")
}

library(survival)
library(survminer)
library(ggplot2)

```

```{r}
#data processing: removing NAs, cleaning "Metastatic Status", and creating "survival time" and "death event" columns.
age_NA_mask <- !is.na(rna_clinical$age_at_diagnosis)
cleaned_rna_clinical <- rna_clinical[age_NA_mask,]

definition_NA_mask <- !is.na(rna_clinical$definition)
cleaned_rna_clinical <- rna_clinical[definition_NA_mask,]

cleaned_rna_clinical$metastatic_status <- ifelse(cleaned_rna_clinical$definition == "Metastatic", "Metastatic", ifelse(cleaned_rna_clinical$definition == "Primary solid Tumor", "Non-metastatic", NA) )

cleaned_rna_clinical$survival_time <- ifelse(cleaned_rna_clinical$vital_status == 'Alive', cleaned_rna_clinical$days_to_last_follow_up, cleaned_rna_clinical$days_to_death)

cleaned_rna_clinical$survival_time <- as.numeric(cleaned_rna_clinical$survival_time)

cleaned_rna_clinical$death_event <- ifelse(cleaned_rna_clinical$vital_status == 'Alive', F, T)

#creating survminer objects
survival_object <- Surv(time = cleaned_rna_clinical$survival_time,
                        event = cleaned_rna_clinical$death_event)

fit_object <- survfit(survival_object ~ metastatic_status, data = cleaned_rna_clinical)

#creating the KM plot
survplot <- ggsurvplot(fit_object,
                       pval=TRUE,
                       ggtheme = theme(plot.margin = unit(c(1,1,1,1), "cm")),
                       legend = 'right')

KM_plot <- survplot$plot + theme_bw() + theme(axis.title = element_text(size=20), 
                                              axis.text = element_text(size=16),
                                              legend.title = element_text(size=14),
                                              legend.text = element_text(size=12))

KM_plot #THIS IS WEIRD
```

***Overall Mutation Differences***
-> using code from intro to MAF to create a cooncoplot
```{r fig.width=13}
#cleaning maf object: getting rid of NAs and factorizing data
maf_object@clinical.data$definition

maf_object@clinical.data$metastatic_cat <- ifelse(maf_object@clinical.data$definition=="Metastatic", "Metastatic", ifelse(maf_object@clinical.data$definition=="Primary solid Tumor", "Non-metastatic", NA))

maf_object@clinical.data$metastatic_cat

factor(maf_object@clinical.data$metastatic_cat, levels=c("Metastatic", "Non-metastatic"))

na_def_mask <- !is.na(maf_object@clinical.data$metastatic_cat)
maf_object@clinical.data <- maf_object@clinical.data[na_def_mask,]

#subsetting maf into metastatic and not metastatic

meta_mask <- ifelse(maf_object@clinical.data$metastatic_cat == "Metastatic", T, F)
meta_barcodes <- maf_object@clinical.data$Tumor_Sample_Barcode[meta_mask]
meta_maf <- subsetMaf(maf=maf_object,
                      tsb = meta_barcodes)


non_meta_mask <- ifelse(maf_object@clinical.data$metastatic_cat == "Non-metastatic", T, F)
non_meta_barcodes <- maf_object@clinical.data$Tumor_Sample_Barcode[non_meta_mask]
non_meta_maf <- subsetMaf(maf=maf_object,
                      tsb = non_meta_barcodes)

#creating a cooncoplot
coOncoplot(m1 = meta_maf, 
           m2 = non_meta_maf, 
           m1Name = 'Metastatic', 
           m2Name = 'Non-metastatic', 
           borderCol = NA)

oncoplot(meta_maf)
oncoplot(non_meta_maf)

```
***Mutation Differences in a Gene of Interest -- DNAH5***
-> code from intro to Maf
```{r}
DNAH5_maf <- subsetMaf(maf = maf_object,
                       genes = 'DNAH5')

maf_object@clinical.data$DNAH5 <- ifelse(maf_object@clinical.data$Tumor_Sample_Barcode %in% DNAH5_maf@clinical.data$Tumor_Sample_Barcode, T, F)

contig <- table(maf_object@clinical.data$metastatic_cat, maf_object@clinical.data$DNAH5)
mosaicplot(contig,
           main = "DNAH5: Metastatic vs Non-Metastatic")

fisher_test <- fisher.test(contig)

fisher_test
fisher_test$p.value
```

***Coocurrence and Mutual Exclusion***

```{r}

somaticInteractions(maf = meta_maf,
                    genes = NULL,
                    pvalue = c(0.01, 0.05))

somaticInteractions(maf = non_meta_maf,
                    genes = NULL,
                    pvalue = c(0.01, 0.05))
```

***Differential Expression***
-> code from Intro to Transcriptomics
```{r fig.height=10}
#loading and setting up data
BiocManager::install("DESeq2")
library(DESeq2)

rna_clinical <- rna_se@colData
rna_clinical <- as.data.frame(rna_clinical)

rna_genes <- rna_se@rowRanges@elementMetadata
rna_genes <- as.data.frame(rna_genes)

rna_counts <- rna_se@assays@data$unstranded
rna_counts <- as.data.frame(rna_counts)

rownames(rna_genes) <- rna_genes$gene_id

rownames(rna_counts) <- rna_genes$gene_id
colnames(rna_counts) <- rownames(rna_clinical)

#data cleaning, including, one-hot encoding, removing NAs, removing <20 counts, and factorizing

#one-hot encoding
rna_clinical$radiation <- ifelse(rna_clinical$patient %in% rad$bcr_patient_barcode, 1, 0)
                             
chemo_mask <- ifelse(drug$pharmaceutical_therapy_type == "Chemotherapy", T, F) #find patients who have chemo
chemo_drug <- drug[chemo_mask,] #create a new df containing ONLY patients with chemo
rna_clinical$chemo <- ifelse(rna_clinical$patient %in% chemo_drug$bcr_patient_barcode, 1, 0) #use IN to input 1s and 0s

immuno_mask <- ifelse(drug$pharmaceutical_therapy_type == "Immunotherapy", T, F)
immuno_drug <- drug[immuno_mask,]
rna_clinical$immuno <- ifelse(rna_clinical$patient %in% immuno_drug$bcr_patient_barcode, 1, 0)

molecular_mask <- ifelse(drug$pharmaceutical_therapy_type == "Targeted Molecular therapy", T, F)
molecular_drug <- drug[molecular_mask,]
rna_clinical$molecular <- ifelse(rna_clinical$patient %in% molecular_drug$bcr_patient_barcode, 1, 0)

vaccine_mask <- ifelse(drug$pharmaceutical_therapy_type == "Vaccine", T, F)
vaccine_drug <- drug[vaccine_mask,]
rna_clinical$vaccine <- ifelse(rna_clinical$patient %in% vaccine_drug$bcr_patient_barcode, 1, 0)

#removing NAs from met_cat, race, gender, vital status

rna_clinical$metastatic_status <- ifelse(rna_clinical$sample_type == "Metastatic", "Metastatic", ifelse(rna_clinical$sample_type == "Primary Tumor", "Non-metastatic", NA) )

meta_NA_mask <- !is.na(rna_clinical$metastatic_status)
DE_rna_clinical <- rna_clinical[meta_NA_mask,]

vital_NA_mask <- !is.na(DE_rna_clinical$vital_status) #this mask removed nothing
DE_rna_clinical <- DE_rna_clinical[vital_NA_mask,]

race_NA_mask <- !is.na(DE_rna_clinical$race)
DE_rna_clinical <- DE_rna_clinical[race_NA_mask,] #this mask removed nothing
race_NR_mask <- ifelse(DE_rna_clinical$race == "not reported", F, T)
DE_rna_clinical <- DE_rna_clinical[race_NR_mask,]

gender_NA_mask <- !is.na(DE_rna_clinical$gender)
DE_rna_clinical <- DE_rna_clinical[gender_NA_mask,] #this mask removed nothing

rna_counts <- rna_counts[,meta_NA_mask]
rna_counts <- rna_counts[,race_NR_mask]

#removing insufficient counts
less_20_mask <- ifelse(rowSums(rna_counts) > 20, T, F)
rna_counts <- rna_counts[less_20_mask,]
rna_genes <- rna_genes[less_20_mask,]

#factorizing
DE_rna_clinical$metastatic_status <- factor(DE_rna_clinical$metastatic_status)
DE_rna_clinical$vital_status <- factor(DE_rna_clinical$vital_status)
DE_rna_clinical$race <- factor(DE_rna_clinical$race)
DE_rna_clinical$gender <- factor(DE_rna_clinical$gender)

#running DESeq
dds <- DESeqDataSetFromMatrix(countData = rna_counts,
                              colData = DE_rna_clinical,
                              design=~ race + gender + vital_status + chemo + radiation + immuno + vaccine + molecular + metastatic_status)

dds_obj <- DESeq(dds)

resultsNames(dds_obj)

results <- results(dds_obj, format = "DataFrame", contrast = c("metastatic_status", "Metastatic", "Non-metastatic"))

DE_rna_clinical$metastatic_status

results <- data.frame(results)
results$gene_name <- rna_genes$gene_name
results$"-log10(padj)" <- -log10(results$padj)
row.names(results) <- rna_genes$gene_id

#Visualizing Results with a Volcano plot
BiocManager::install("EnhancedVolcano")
library(EnhancedVolcano)

EnhancedVolcano(results,
                lab = results$gene_name,
                x = 'log2FoldChange',
                y = 'pvalue',
                title = 'Metastatic vs Non-Metastatic',
                pointSize = 1.0,
                labSize = 5.0)


```

